[ticket/16851] Add Amazonbot, AhrefsBot and SemrushBot

MichaIng opened the pull request:
Checklist:

- [x] Correct branch: master for new features; 3.3.x for fixes
- [ ] Tests pass
- [x] Code follows coding guidelines: [master](https://area51.phpbb.com/docs/dev/master/development/coding_guidelines.html) and [3.3.x](https://area51.phpbb.com/docs/dev/3.3.x/development/coding_guidelines.html)
- [x] Commit follows commit message [format](https://area51.phpbb.com/docs/dev/3.3.x/development/git.html)

Tracker ticket (set the ticket ID to **your ticket ID**):

https://tracker.phpbb.com/browse/PHPBB3-16851
______________
The three bot lists have further been aligned the following way:
- Complement each list with all bots present in another list, in case. There might be a reason why a bot has been removed from a list, but then it should be removed from all list, for consistency.
- Remove a bots version integer from user agent match string, if present in one list but missing in another. Generally, since versions change, it makes sense to match it, so the entry has a chance to remain valid over a longer time.
- Add trailing slash to user agent match string, if missing in one list but present in another. Nearly all bots have a slash in their user agent name, separating the bots name from the version string. Including it into the match makes sense to lower the chance of false positives a little.
- Remove `+(` from Baidu spider, which is nowadays not part of its user agent. Instead, there are many different Baidu spiders, with or without a dash + content type prefixed. The used string matches all of them: https://qpsoftware.net/blog/baiduspider
- Add the wildcard to all W3C(*)_Validator strings, although it doesn't seem to be required nowadays: http://validator.w3.org/services

More information about the added/changed bots:
- Amazonbot: https://developer.amazon.com/de/support/amazonbot
- AhrefsBot: https://ahrefs.com/robot
- SemrushBot: https://www.semrush.com/bot/
- Baiduspider: https://qpsoftware.net/blog/baiduspider
- Exabot: https://www.webrobots.de/exabot/
- W3C_Validator: http://validator.w3.org/services

DavidIQ commented:
There's no migration here so these would never get added to an existing installation.

MichaIng commented:
Thanks for the hint. I wasn't sure if new bots are supposed to be added to the list without admins doing an explicit repair steps or so. I'll check our how such migration was done in other cases and will add it to the commit.

3D-I commented:
See for reference: https://github.com/phpbb/phpbb/pull/5919

MichaIng commented:
I added the additional bots into the existing migration script. It loops through all new bots provided as a name => agent array and then keeps using the existing code to add them one by one. I removed the `bot_ip` variable, as for the default bots (running from several different IPs) this is not used anymore, and hardcoded it as empty string in the SQL query, I hope this is okay.

3D-I reviewed (changes requested):
You need to create a brand new migration and keep the old ones untouched.

3D-I commented on phpBB/phpbb/db/migration/data/v33x/bot_update.php:
```diff
@@ -23,26 +23,32 @@ static public function depends_on()
+			$bot_exists = (bool) $this->db->sql_fetchfield('user_id');
+			$this->db->sql_freeresult($result);
+
+			if ($bot_exists) continue;
```
That's not allowed.

3D-I replied:
That's not allowed.

MichaIng commented:
`if ($bot_exists) continue;` seems to be not allowed by the CI. I'll make a regular multi line block of this.

> You need to create a brand new migration and keep the old ones untouched.

Ah okay, sad as this means a lot of doubled code. Why is this necessary? Don't all of those migrations in `3.3.x` on each update of those minor versions? Or is it stored somewhere that this migration was running already and is skipped then?

3D-I commented:
Ofcourse migrations are recorded already.

MichaIng commented:
Probably we can create a general bot update migration script that is repeatedly running. Although it would mean, if an admin removed a bot actively, the next update would re-add it. So it makes sense to consequently run it a single time only, to prevent that...

3D-I commented:
I have a bot update sript already made. But no, that's not the way it goes here.

3D-I commented on phpBB/phpbb/db/migration/data/v33x/bot_update_vol2.php:
```diff
@@ -0,0 +1,99 @@
+{
+	static public function depends_on()
+	{
+		return ['\phpbb\db\migration\data\v330\v330'];
+	}
```
`return ['\phpbb\db\migration\data\v33x\v334'];`

MichaIng replied:
It does not "depend" on the migration step done before in v3.3.x, but probably good to have things in a clear order. Change applied.

3D-I commented on phpBB/phpbb/db/migration/data/v33x/bot_update_vol2.php:
```diff
@@ -0,0 +1,99 @@
+{
+	static public function depends_on()
+	{
+		return ['\phpbb\db\migration\data\v334\v334'];
```
`return ['\phpbb\db\migration\data\v33x\v334'];`

MichaIng replied:
Ah dammit, thanks.

MichaIng commented:
A bot migration function could be also added to core, so that the individual migration scripts only need to call that single function, like `add_new_bot($bot_name, $bot_agent, $bot_ip);`. But on the other hand, new bots are probably not added regular enough to bother with this.

MichaIng commented on phpBB/phpbb/db/migration/data/v33x/bot_update_vol2.php:
```diff
@@ -0,0 +1,99 @@
+		$bots = [
+			'Ahrefs [Bot]' => 'AhrefsBot/',
+			'Amazon [Bot]' => 'Amazonbot/',
+			'Semrush [Bot]' => 'SemrushBot/'
```
I see the other arrays have a trailing comma for the last item as well. Shall I add this, given the tests do not fail on it anyway?

3D-I replied:
I do not think is necessary, probably though by the coding guidelines... I do not recall now.

marc1706 replied:
Please add it.

3D-I commented:
> 
> 
> A bot migration function could be also added to core, so that the individual migration scripts only need to call that single function, like `add_new_bot($bot_name, $bot_agent, $bot_ip);`. But on the other hand, new bots are probably not added regular enough to bother with this.

I do not think is necessary, the bots you are adding I added to my board since years, in ACP.
From the whole list of bots only 40% of them visited my board.

![image](https://user-images.githubusercontent.com/480857/130353082-90abba99-5d75-44e6-9324-954d64278955.png)

MichaIng commented:
> From the whole list of bots only 40% of them visited my board.

Same here, but which bots are the active ones seems to depend on the individual website, probably backlinks, robot tags/headers etc. On our forum, the three added bots are the most active ones, with several thousand requests every day, most of all with large distance the Amazonbot:
- Amazonbot: ~120,000 requests every day (!)
- AhrefsBot: ~8,000 requests every day
- SemrushBot: ~3,000 - ~9,000 (quite an increase today)
- Googlebot: ~1,500, and I'm quite actively maintaining the domain on Google's search console.

marc1706 commented on phpBB/phpbb/db/migration/data/v33x/bot_update_vol2.php:
```diff
@@ -0,0 +1,99 @@
+
+class bot_update_vol2 extends \phpbb\db\migration\migration
+{
+	static public function depends_on()
```
```suggestion
	public static function depends_on()
```

MichaIng replied:
Shall I adjust this in the old bot update script as well?

marc1706 commented on phpBB/phpbb/db/migration/data/v33x/bot_update_vol2.php:
```diff
@@ -0,0 +1,99 @@
+
+namespace phpbb\db\migration\data\v33x;
+
+class bot_update_vol2 extends \phpbb\db\migration\migration
```
```suggestion
class bot_update_v2 extends \phpbb\db\migration\migration
```

marc1706 commented on phpBB/phpbb/db/migration/data/v33x/bot_update_vol2.php:
```diff
@@ -0,0 +1,99 @@
+			'Amazon [Bot]' => 'Amazonbot/',
+			'Semrush [Bot]' => 'SemrushBot/'
+		];
+
```
```suggestion
		$group_row = [];
```

marc1706 commented on phpBB/phpbb/db/migration/data/v33x/bot_update_vol2.php:
```diff
@@ -0,0 +1,99 @@
+				continue;
+			}
+
+			$sql = 'SELECT group_id, group_colour
+				FROM ' . $this->table_prefix . 'groups
+				WHERE ' . $this->db->sql_build_array('SELECT', ['group_name' => 'BOTS']);
+			$result = $this->db->sql_query($sql);
+			$group_row = $this->db->sql_fetchrow($result);
+			$this->db->sql_freeresult($result);
+
+			// Default fallback, should never get here
+			if (!$group_row)
+			{
+				$group_row['group_id'] = 6;
+				$group_row['group_colour'] = '9E8DA7';
+			}
```
```suggestion
			if (!count($group_row))
			{
				$sql = 'SELECT group_id, group_colour
					FROM ' . $this->table_prefix . 'groups
					WHERE ' . $this->db->sql_build_array('SELECT', ['group_name' => 'BOTS']);
				$result = $this->db->sql_query($sql);
				$group_row = $this->db->sql_fetchrow($result);
				$this->db->sql_freeresult($result);

				// Default fallback, should never get here
				if (!count($group_row))
				{
					$group_row['group_id'] = 6;
					$group_row['group_colour'] = '9E8DA7';
				}
			}
```

marc1706 replied:
With this and the above change we won't have to re-run the same query three times.

marc1706 commented on phpBB/phpbb/db/migration/data/v33x/bot_update_vol2.php:
```diff
@@ -0,0 +1,99 @@
+			$user_id = user_add($user_row);
+			$sql = 'INSERT INTO ' . $this->table_prefix . 'bots ' . $this->db->sql_build_array('INSERT', [
+				'bot_active'	=> 1,
+				'bot_name'		=> (string) $bot_name,
+				'user_id'		=> (int) $user_id,
+				'bot_agent'		=> (string) $bot_agent,
```
```suggestion
				'bot_name'		=> $bot_name,
				'user_id'		=> (int) $user_id,
				'bot_agent'		=> $bot_agent,
```

MichaIng replied:
Shall I adjust this in the old bot update script as well?

marc1706 replied:
No, no point in changing the old migration.
